### PR TITLE
FromDDL rework

### PR DESCRIPTION
**Pull request by svaningelgem**
I basically took the tests from spark itself, and wrote a lex-parser around it (as I couldn't find an easy way to extract the one from scala).
It supports every test that was in the scala program, including:
- basic datatypes
- maps
- structs
- arrays
- weird names (like `a+b`)
- comments
- reserved names-names (like `int: int` -- int is both the name of the column, and the datatype! -- But "int int" doesn't work the same way, that errors out.)

@tools4origins : could you check if this can be integrated into your PR because I believe my implementation is more feature complete?

### CODE REVIEW

File: pysparkling/tests/test_fromDDL.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+Most of these test cases come from pyspark.sql.types._parse_datatype_string()
+"""
+import pytest
+
+from pysparkling.sql.types import (
+    ArrayType, ByteType, DateType, DecimalType, DoubleType, IntegerType, LongType, MapType, ShortType, StringType,
+    StructField, StructType
+)
+# DataType
+from pysparkling.sql.utils import ParseException
```
Comment (tools4origins) [inline pysparkling/tests/test_fromDDL.py:11]: StructType does not expose fromDDL in Python, (maybe we should not either), but the one in Scala does not show the behavior expected in these tests:

```scala
scala> StructType.fromDDL("some_str string, some_int integer, some_date date")
res2: org.apache.spark.sql.types.StructType = StructType(StructField(some_str,StringType,true), StructField(some_int,IntegerType,true), StructField(some_date,DateType,true))
```

```scala
scala> StructType.fromDDL("int")
org.apache.spark.sql.catalyst.parser.ParseException:
mismatched input '<EOF>'
```

(Spark 3)

Comment (tools4origins) [reply pysparkling/tests/test_fromDDL.py:11]: Multiple types of syntaxes are supported, fromDDL only handles one of them in Spark:
https://github.com/apache/spark/blob/master/python/pyspark/sql/types.py#L821-L841

Maybe we want not to rely on StructType.fromDDL to do this parsing?

Comment (svaningelgem) [reply pysparkling/tests/test_fromDDL.py:11]: I think this method is wrongly called. We shouldn't call this fromDDL. This is actually `_parse_datatype_string`. And it's supported by Spark ;-). This is mainly used when you create a dataframe with a string as the field description.

Just check the implementation in pyspark, it supports multiple ways and I do agree that we should drop fromDDL (as it's not in pyspark at all).

File: pysparkling/sql/tests/test_types_parser.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+# checkDataType("varchAr(20)", VarcharType(20))
+# checkDataType("cHaR(27)", CharType(27))
+
+
+def test_BINARY():
+    assert parser.parse("BINARY") == BinaryType()
+
+
+def test_void():
+    assert parser.parse("void") == NullType()
```
Comment (tools4origins) [inline pysparkling/sql/tests/test_types_parser.py:118]: Where is this from?

```python
>>> spark.createDataFrame([], "id void")
[...[
pyspark.sql.utils.ParseException: 
DataType void is not supported.(line 1, pos 3)

== SQL ==
id void
---^^^
```

Comment (svaningelgem) [reply pysparkling/sql/tests/test_types_parser.py:118]: https://github.com/apache/spark/blob/54c053afb0c9d3fcc7ac311100c8db9deeb163c0/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DataTypeParserSuite.scala#L64

**Pull request closed without merging**